source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1037,7 +1038,193 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::httpAsyncClient
+  return OdCdsApiHandleImpl::create(*this, std::move(odcds));
+}
+
+ClusterDiscoveryCallbackHandlePtr ClusterManagerImpl::requestOnDemandClusterDiscovery(
+    OdCdsApiSharedPtr odcds, const std::string& name, ClusterDiscoveryCallbackSharedPtr callback,
+    std::chrono::milliseconds timeout) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, callback);
+  // This check will catch requests for discoveries from this thread only. If other thread requested
+  // the same discovery, we will detect it in the main thread later.
+  if (discovery_in_progress) {
+    ENVOY_LOG(debug,
+              "cm odcds: on-demand discovery for cluster {} is already in progress, something else "
+              "in thread {} has already requested it",
+              name, cluster_manager.thread_local_dispatcher_.name());
+    // This worker thread has already requested a discovery of a cluster with this name, so nothing
+    // more left to do here.
+    return std::move(handle);
+  }
+  ENVOY_LOG(
+      debug,
+      "cm odcds: forwarding the on-demand discovery request for cluster {} to the main thread",
+      name);
+  // This seems to be the first request for discovery of this cluster in this worker thread. Rest of
+  // the process may only happen in the main thread.
+  dispatcher_.post([this, odcds = std::move(odcds), timeout,
+                    weak_callback = ClusterDiscoveryCallbackWeakPtr(callback), name,
+                    &thread_local_dispatcher = cluster_manager.thread_local_dispatcher_] {
+    // Check for the cluster here too. It might have been added between the time when this closure
+    // was posted and when it is being executed.
+    if (getThreadLocalCluster(name) != nullptr) {
+      ENVOY_LOG(
+          debug,
+          "cm odcds: the requested cluster {} is already known, posting the callback back to {}",
+          name, thread_local_dispatcher.name());
+      thread_local_dispatcher.post([name, weak_callback, &thread_local_dispatcher] {
+        if (auto callback = weak_callback.lock(); callback != nullptr) {
+          ENVOY_LOG(debug, "cm odcds: invoking callback on the already known cluster {} in {}",
+                    name, thread_local_dispatcher.name());
+          // If this gets called here, it means that we requested a discovery of a cluster without
+          // checking if that cluster is already known by cluster manager. The other situation we
+          // get here can be when another thread was faster in requesting the discovery and handling
+          // the resulting new cluster. It's a rather unlikely race, though.
+          (*callback)(ClusterDiscoveryStatus::Available);
+        }
+      });
+      return;
+    }
+
+    if (auto it = pending_cluster_creations_.find(name); it != pending_cluster_creations_.end()) {
+      ENVOY_LOG(debug, "cm odcds: on-demand discovery for cluster {} is already in progress", name);
```

Comment:
Coverage is generally pretty good, but there are some missing spots, e.g. https://storage.googleapis.com/envoy-pr/722562c/coverage/source/common/upstream/cluster_manager_impl.cc.gcov.html. Can you add any red that appears in the code that you have added?